python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -302,55 +302,69 @@ def execute_asset_backfill_iteration_inner(
     if request_roots:
         initial_candidates.update(asset_backfill_data.get_target_root_asset_partitions())
 
-    (
-        parent_materialized_asset_partitions,
-        next_latest_storage_id,
-    ) = find_parent_materialized_asset_partitions(
-        asset_graph=asset_graph,
-        instance_queryer=instance_queryer,
-        target_asset_selection=AssetSelection.keys(*asset_backfill_data.target_subset.asset_keys),
-        latest_storage_id=asset_backfill_data.latest_storage_id,
-    )
-    initial_candidates.update(parent_materialized_asset_partitions)
-
-    yield None
-
-    recently_materialized_asset_partitions = AssetGraphSubset(asset_graph)
-    for asset_key in asset_backfill_data.target_subset.asset_keys:
-        records = instance_queryer.get_materialization_records(
-            asset_key=asset_key,
-            after_cursor=asset_backfill_data.latest_storage_id,
-            tags=frozendict({BACKFILL_ID_TAG: backfill_id}),
+        next_latest_storage_id = instance_queryer.get_latest_storage_id(
+            DagsterEventType.ASSET_MATERIALIZATION
         )
-        recently_materialized_asset_partitions |= {
-            AssetKeyPartitionKey(asset_key, record.partition_key) for record in records
-        }
+        updated_materialized_subset = AssetGraphSubset(asset_graph)
+        failed_and_downstream_subset = AssetGraphSubset(asset_graph)
+    else:
+        (
+            parent_materialized_asset_partitions,
+            next_latest_storage_id,
+        ) = find_parent_materialized_asset_partitions(
+            asset_graph=asset_graph,
+            instance_queryer=instance_queryer,
+            target_asset_selection=AssetSelection.keys(
+                *asset_backfill_data.target_subset.asset_keys
+            ),
+            latest_storage_id=asset_backfill_data.latest_storage_id,
+        )
+        initial_candidates.update(parent_materialized_asset_partitions)
 
         yield None
 
-    updated_materialized_asset_partitions = (
-        asset_backfill_data.materialized_subset | recently_materialized_asset_partitions
-    )
+        recently_materialized_asset_partitions = AssetGraphSubset(asset_graph)
+        for asset_key in asset_backfill_data.target_subset.asset_keys:
+            records = instance_queryer.get_materialization_records(
+                asset_key=asset_key, after_cursor=asset_backfill_data.latest_storage_id
+            )
+            records_in_backfill = [
+                record
+                for record in records
+                if instance_queryer.run_has_tag(
```

Comment:
Is it possible to pull this query out of the loop and batch?  Or limit the number of repeated calls to fetch run tags in case the run_id is repeated (e.g. materialize multiple assets within the same run)?

Reply:
the instance_queryer handles caching (in `_get_run_record_by_id`)

Reply:
ah, missed the cached_method decorator.